nat-lab/tests/test_events_link_state.py

```diff
@@ -0,0 +1,489 @@
+import asyncio
+import pytest
+from contextlib import AsyncExitStack
+from helpers import SetupParameters, setup_mesh_nodes
+from telio import AdapterType, LinkStateEventsCount
+from telio_features import (
+    TelioFeatures,
+    NoLinkDetection,
+    Wireguard,
+    PersistentKeepalive,
+)
+from utils import testing
+from utils.connection_tracker import ConnectionLimits
+from utils.connection_util import generate_connection_tracker_config, ConnectionTag
+from utils.ping import Ping
+
+
+def long_persistent_keepalive_periods() -> Wireguard:
+    return Wireguard(
+        persistent_keepalive=PersistentKeepalive(
+            vpn=3600, direct=3600, proxying=3600, stun=3600
+        )
+    )
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "alpha_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_1,
+                adapter_type=AdapterType.LinuxNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_1,
+                    derp_1_limits=ConnectionLimits(1, 1),
+                ),
+                features=TelioFeatures(
+                    no_link_detection=NoLinkDetection(rtt_seconds=1),
+                    wireguard=long_persistent_keepalive_periods(),
+                ),
+            )
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "beta_setup_params",
+    [
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.DOCKER_CONE_CLIENT_2,
+                adapter_type=AdapterType.LinuxNativeWg,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.DOCKER_CONE_CLIENT_2,
+                    derp_1_limits=ConnectionLimits(1, 1),
+                ),
+                features=TelioFeatures(
+                    no_link_detection=NoLinkDetection(rtt_seconds=1),
+                    wireguard=long_persistent_keepalive_periods(),
+                ),
+            )
+        )
+    ],
+)
+async def test_event_link_state_peers_idle_all_time(
+    alpha_setup_params: SetupParameters, beta_setup_params: SetupParameters
+) -> None:
+    async with AsyncExitStack() as exit_stack:
+        env = await setup_mesh_nodes(
+            exit_stack, [alpha_setup_params, beta_setup_params]
+        )
+        alpha, beta = env.nodes
+        client_alpha, client_beta = env.clients
+
+        # Expect no link event while peers are idle
+        await asyncio.sleep(20)
+        alpha_counts = client_beta.get_link_state_events_count(alpha.public_key)
+        beta_counts = client_alpha.get_link_state_events_count(beta.public_key)
+
+        # 1 down when node is Connecting and 1 up when node is Connected
```

Comment:
piggy backing on the previous comment about the order - due to a bug it might be  that events come in as `up` and then `down` yet this test would pass.